lib/saluki-config/src/refresher.rs

```diff
@@ -0,0 +1,131 @@
+use std::sync::Arc;
+
+use arc_swap::ArcSwap;
+use reqwest::ClientBuilder;
+use saluki_error::GenericError;
+use serde::{de::DeserializeOwned, Deserialize};
+use serde_json::Value;
+use tokio::time::{sleep, Duration};
+use tracing::{debug, error};
+
+use crate::{ConfigurationError, GenericConfiguration};
+
+const DATADOG_AGENT_CONFIG_ENDPOINT: &str = "https://localhost:5004/config/v1/";
+const DEFAULT_AUTH_TOKEN_FILE_PATH: &str = "/etc/datadog-agent/auth_token";
+const DEFAULT_REFRESH_INTERVAL_SECONDS: u64 = 15;
+
+/// Configuration for setting up `RefreshableConfiguration`.
+#[derive(Default, Deserialize)]
+pub struct RefresherConfiguration {
+    /// The location of the auth token used by the datadog agent.
+    ///
+    /// Defaults to `/etc/datadog-agent/auth_token`.`
+    #[serde(default = "default_auth_token_file_path")]
+    auth_token_file_path: String,
+
+    /// The amount of time betweeen each request in seconds.
+    ///
+    /// Defaults to 15 seconds.
+    #[serde(default = "default_refresh_interval_seconds")]
+    refresh_interval_seconds: u64,
+}
+
+fn default_auth_token_file_path() -> String {
+    DEFAULT_AUTH_TOKEN_FILE_PATH.to_owned()
+}
+
+fn default_refresh_interval_seconds() -> u64 {
+    DEFAULT_REFRESH_INTERVAL_SECONDS
+}
+
+/// A configuration whose values are refreshed from a remote source at runtime.
+#[derive(Clone, Default)]
+pub struct RefreshableConfiguration {
+    token: String,
+    values: Arc<ArcSwap<Value>>,
+    refresh_interval_seconds: u64,
+}
+
+impl RefresherConfiguration {
+    /// Creates a new `RefresherConfiguration` from the given configuration.
+    pub fn from_configuration(config: &GenericConfiguration) -> Result<Self, GenericError> {
+        Ok(config.as_typed()?)
+    }
+    /// Create `RefreshableConfiguration` from `RefresherConfiguration`.
+    pub fn build(&self) -> Result<Arc<RefreshableConfiguration>, GenericError> {
+        let raw_bearer_token = std::fs::read_to_string(&self.auth_token_file_path)?;
+        let refreshable_configuration = Arc::new(RefreshableConfiguration {
```

Comment:
We don't need to wrap `RefreshableConfiguration` in `Arc`, since the internal values state is already wrapped in `Arc`.